adafruit_irremote.py

```diff
@@ -69,118 +66,219 @@ class IRNECRepeatException(Exception):
-            ]
-        # convert marks/spaces to 0 and 1
-        for i, pulse_length in enumerate(pulses):
-            if (space * 0.75) <= pulse_length <= (space * 1.25):
-                pulses[i] = False
-            elif (mark * 0.75) <= pulse_length <= (mark * 1.25):
-                pulses[i] = True
-            else:
-                raise IRDecodeException("Pulses outside mark/space")
-
-        # convert bits to bytes!
-        output = [0] * ((len(pulses) + 7) // 8)
-        for i, pulse_length in enumerate(pulses):
-            output[i // 8] = output[i // 8] << 1
-            if pulse_length:
-                output[i // 8] |= 1
-        return output
+
+class NonblockingGenericDecode:
+    """
+    Decode pulses into bytes in a non-blocking fashion.
+
+    :param ~pulseio.PulseIn input_pulses: Object to read pulses from
+    :param int max_pulse: Pulse duration to end a burst.  Units are microseconds.
+
+    >>> pulses = PulseIn(...)
+    >>> decoder = NonblockingGenericDecoder(pulses)
+    >>> for message in decoder.read():
+    ...     if isinstace(message, IRMessage):
+    ...         message.code  # TA-DA! Do something with this in your application.
+    ...     else:
+    ...         # message is either NECRepeatIRMessage or
+    ...         # UnparseableIRMessage. You may decide to ignore it, raise
+    ...         # an error, or log the issue to a file. If you raise or log,
+    ...         # it may be helpful to include message.pulses in the error message.
+    ...         ...
+    """
+
+    def __init__(self, pulses, max_pulse=10_000):
+        self.pulses = pulses  # PulseIn
+        self.max_pulse = max_pulse
+        self._unparsed_pulses = []  # internal buffer of partial messages
+
+    def read(self):
+        """
+        Consume all pulses from PulseIn. Yield decoded messages, if any.
+
+        If a partial message is received, this does not block to wait for the
+        rest. It stashes the partial message, to be continued the next time it
+        is called.
+        """
+        # Consume from PulseIn.
+        while self.pulses:
+            pulse = self.pulses.popleft()
+            self._unparsed_pulses.append(pulse)
+            if pulse > self.max_pulse:
+                # End of message! Decode it and yield a BaseIRMessage.
+                try:
+                    yield decode_bits(self._unparsed_pulses)
+                except FailedToDecode as err:
+                    # If you want to debug failed decodes, this would be a good
+                    # place to print/log or (re-)raise.
+                    (unparseable_message,) = err.args
+                    yield unparseable_message
+                self._unparsed_pulses.clear()
+                # TODO Do we need to consume and throw away more pulses here?
+                # I'm unclear about the role that "pruning" plays in the
+                # original implementation in GenericDecode._read_pulses_non_blocking.
```

Comment:
I'm not sure either. I think we can do without it now and tweak later if it doesn't work for folks.